debugger/vscode-dap-extension/src/starlingMonkeyRuntime.ts

```diff
@@ -0,0 +1,677 @@
+        }
+      });
+    } else {
+      // Fallback to sendText if there is no shell integration.
+      // Send Ctrl+C to kill any existing component runtime first.
+      this._terminal!.sendText('\x03', false);
+      this._terminal!.sendText(
+        `${config.componentRuntime.executable} ${componentRuntimeArgs.join(" ")}`,
+        true
+      );
+    }
+  }
+
+  private static async ensureTerminal() {
+    if (this._terminal && this._terminal.exitStatus === undefined) {
+      return;
+    }
+
+    const signal = new Signal<void, void>();
+    this._terminal = window.createTerminal();
+    const terminalCloseDisposable = window.onDidCloseTerminal((terminal) => {
+      if (terminal === this._terminal) {
+        signal.resolve();
+        this._terminal = undefined;
+        this._runtimeExecution = undefined;
+        terminalCloseDisposable.dispose();
+      }
+    });
+
+    const shellIntegrationDisposable = window.onDidChangeTerminalShellIntegration(
+      async ({ terminal }) => {
+        if (terminal === this._terminal) {
+          clearTimeout(timeout);
+          shellIntegrationDisposable.dispose();
+          signal.resolve();
+        }
+      }
+    );
+    // Fallback to sendText if there is no shell integration within 3 seconds of launching
+    const timeout = setTimeout(() => {
+      shellIntegrationDisposable.dispose();
+      signal.resolve();
+    }, 3000);
+
+    await signal.wait();
+  }
+}
+
+export class StarlingMonkeyRuntime extends EventEmitter<RuntimeEventMap> {
+  private _debug!: boolean;
+  private _stopOnEntry!: boolean;
+  private _sourceMaps!: SourceMaps;
+  public get fileAccessor(): FileAccessor {
+    return this._fileAccessor;
+  }
+  public set fileAccessor(value: FileAccessor) {
+    this._fileAccessor = value;
+  }
+
+  private _server!: Net.Server;
+  private _socket!: Net.Socket;
+
+  private _state: RuntimeState = { state: 'init' };
+  private _messageReceived = new Signal<InstanceToRuntimeMessage, void>();
+
+  private _sourceFile!: string;
+  public get sourceFile() {
+    return this._sourceFile;
+  }
+
+  constructor(
+    private _workspaceDir: string,
+    private _fileAccessor: FileAccessor,
+    private _config: IStarlingMonkeyRuntimeConfig
+  ) {
+    super();
+  }
+
+  public async start(program: string, component: string, stopOnEntry: boolean, debug: boolean  ): Promise<void> {
+    this._state = { state: 'connecting' };
+    await ComponentRuntimeInstance.start(this._workspaceDir, component, this._config);
+    this.startSessionServer();
+    // TODO: tell StarlingMonkey not to debug if this is false.
+    this._debug = debug;
+    this._stopOnEntry = stopOnEntry;
+    this._sourceFile = this.normalizePath(program);
+
+    this.runDebugLoop();
+  }
+
+  async runDebugLoop() {
+    // We have a slightly odd mix here of state machine stuff (unsolicited messages)
+    // and request-response stuff.  Feels like there should be a better way.
+
+    while (true) {
```

Comment:
It's unclear how we exit this loop. Should this break on some `message.type`?

Reply:
`"end"` is a JS event, not a message; it's handled at `starlingMonkeyDebugger.ts` line 123.  (The events and messages thing is a bit fiddly, especially coming in cold, because there's a lot of overlap.  I tried to tease them apart by declaring the RuntimeEventMap (events) and the RuntimeToInstanceMessage/InstanceToRuntimeMessage (socket messages) but it still doesn't jump out to the eye I realise.)